Documentation/proposals/202310-automated-sharding.md

```diff
@@ -0,0 +1,130 @@
+- Prometheus Servers: `rate(prometheus_tsdb_head_samples_appended_total[5m])`
+
+3. Or even measuring samples queried per second for Prometheus instances under heavy query load.
+
+- Prometheus Servers: `rate(prometheus_engine_query_samples_total[5m])`
+
+Modern Custom Horizontal Pod Autoscalers (HPAs), such as Keda are already capable of [scaling CRDs](https://keda.sh/docs/2.11/concepts/scaling-deployments/#scaling-of-custom-resources) based on different parameters, such as [CPU, Memory](https://keda.sh/docs/2.11/scalers/cpu/) or even [custom Prometheus metrics](https://keda.sh/docs/2.11/scalers/prometheus/).
+
+### Scale Status subresource
+
+When working with CRDs, common HPAs (e.g. Keda) depends on a Status subresource called [Scale](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource). Unfortunately, Prometheus-Operator resources still don't implement this field. There was an attempt to implement this in https://github.com/prometheus-operator/prometheus-operator/pull/4735, but was reverted because the PR was scaling replicas instead of shards. Although scaling replicas can help with High-Availability, it is expensive and hard to manage since it duplicates scrapes while not reducing the load on top of Prometheus because all replicas still scrape the same targets. Sharding serves as a better autoscaling strategy since the load is spread into all instances instead of getting duplicated.
+
+With only this change, Prometheus Agents can already be horizontally scaled without problems, but for Prometheus Servers it gets a little more complicated.
+
+### Graceful scale down of Prometheus Servers
+
+Prometheus Servers don't behave like the usual stateless applications. A few cases require Prometheus pods to still be available even after a scale down event is triggered:
+
+* Thanos proxying queries back to the sidecars to retrieve data that still wasn't uploaded to object storage.
+* Prometheus servers with long retention periods still hold data that users might want to query.
+
+If we simply delete Prometheus pods during scale-down, data loss will be a common problem, so a more complex scale-down strategy needs to be used here.
+
+We propose that Prometheus-Operator introduces a new field to the Prometheus Server CRDs:
+
+```yaml
+apiVersion: monitoring.coreos.com
+kind: Prometheus
+metadata:
+  name: example
+spec:
+  autoscaling:
+    scaleDownStrategy: force|preventDataLoss # Default: force
+```
+
+If scaleDownStrategy == force, the operator would simply delete the pod on the spot, not caring about the data stored in that particular shard. Although not the safest approach, it is exactly the current behavior nowadays and changing it might surprise a lot of users.
+
+If scaleDownStrategy == preventDataLoss then Prometheus-Operator would first mark the pod with an annotation:
+
+```yaml
+prometheus.operator.io/safe-deletion-timestamp: X
```

Comment:
We could get some inspiration on kubectl and have something like `grace-period` this is the name of a existing flag when running kubectl delete pod.

`prometheus.operator.io/grace-period`

wdyt?

Reply:
It's a bit different, grace period is a duration, while here we set an explicit Unix timestamp.

I'm still open for a better name, I just think that it should make it clear that it is indeed a timestamp and not a duration

Reply:
Hmmm 🤔 
Why not a duration, i think it would be even more clear.
prometheus.operator.io/grace-period: 2h

Means Prometheus Operator will wait 2h after scale down signal to delete this instance.

No?

Reply:
what if Prometheus-Operator restarts? will it wait for another 2 hours?

Reply:
We need in someway keep the state.
Because the timestamp will be different from time to time right.
I mean users will need to update the timestamp.
Or I'm missing something?

Reply:
I don't think a state is needed here, the implementation is much more complex.

We can have a Unix timestamp and if current time > annotation then we trigger the deletion 🤔 

Imagine the following situations:

* When setting the annotation:

```go
retention := prom.args["--storage.tsdb.retention.time"]
prom.annotations["prometheus.operator.io/safe-deletion-timestamp"] = time.Now().Add(retention).Unix()
```

* When deciding if it needs to be deleted
```go
deletionTs := time.Parse(prom.annotations["prometheus.operator.io/safe-deletion-timestamp"])

if time.Now() > deletionTs {
  // delete prom
}
```

--- 
Of course the code here is very superficial, but the logic would be something like that

Reply:
Maybe I miss understand this.
Users will not provide that annotation, we're going to do all the automation.
So it will be always calculated by the Operator.

is that right?

Reply:
> Maybe I miss understand this. Users will not provide that annotation, we're going to do all the automation. So it will be always calculated by the Operator.
> 
> is that right?

correct!

Reply:
Ok!
I got it.
I could suggest still prometheus.operator.io/grace-period-timestamp 😄

Reply:
```suggestion
operator.prometheus.io/safe-deletion-timestamp: X
```
Regarding the naming, grace-period feels slightly off to me as I would expect that to be a duration. How about "scheduled-deletion-timestamp"? This would mark the intent better imo.

Reply:
Seems cool also

Reply:
Agreed that it should a point in time in the future and not a period.

(bike-shedding) what about `operator.prometheus.io/deletion-timestamp: 2023-10-05T08:18:09+00:00Z`?